dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java

```diff
@@ -172,63 +169,118 @@ private void handleDataValues(
+    EventDataValue eventDataValue = new EventDataValue();
+    eventDataValue.setDataElement(dataElement.getUid());
+    eventDataValue.setCreated(new Date());
+    eventDataValue.setCreatedByUserInfo(UserInfoSnapshot.from(user));
+    eventDataValue.setStoredBy(user.getUsername());
 
-          eventDataValue.setDataElement(dataElement.getUid());
-          eventDataValue.setStoredBy(dv.getStoredBy());
+    eventDataValue.setLastUpdated(new Date());
+    eventDataValue.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
 
-          if (StringUtils.isEmpty(dv.getValue())) {
-            if (dataElement.isFileType()) {
-              unassignFileResource(
-                  entityManager, preheat, event.getUid(), eventDataValue.getValue());
-            }
+    eventDataValue.setValue(dv.getValue());
+    eventDataValue.setProvidedElsewhere(dv.isProvidedElsewhere());
 
-            event.getEventDataValues().remove(eventDataValue);
-          } else {
-            eventDataValue.setValue(dv.getValue());
+    if (dataElement.isFileType()) {
+      assignFileResource(entityManager, preheat, event.getUid(), eventDataValue.getValue());
+    }
 
-            if (dataElement.isFileType()) {
-              assignFileResource(entityManager, preheat, event.getUid(), eventDataValue.getValue());
-            }
+    event.getEventDataValues().add(eventDataValue);
+  }
 
-            event.getEventDataValues().remove(eventDataValue);
-            event.getEventDataValues().add(eventDataValue);
-          }
+  private void updateDataValue(
+      EventDataValue eventDataValue,
+      DataValue dv,
+      Event event,
+      DataElement dataElement,
+      UserDetails user,
+      EntityManager entityManager,
+      TrackerPreheat preheat) {
+    eventDataValue.setLastUpdated(new Date());
+    eventDataValue.setLastUpdatedByUserInfo(UserInfoSnapshot.from(user));
 
-          logTrackedEntityDataValueHistory(
-              user.getUsername(), dataElement, event, new Date(), valuesHolder);
-        });
+    if (dataElement.isFileType()) {
+      unassignFileResource(entityManager, preheat, event.getUid(), eventDataValue.getValue());
+      assignFileResource(entityManager, preheat, event.getUid(), dv.getValue());
+    }
+
+    eventDataValue.setProvidedElsewhere(dv.isProvidedElsewhere());
+    eventDataValue.setValue(dv.getValue());
   }
 
-  private Date getFromOrNewDate(DataValue dv, Function<DataValue, Instant> dateGetter) {
-    return Optional.of(dv).map(dateGetter).map(DateUtils::fromInstant).orElseGet(Date::new);
+  private void deleteDataValue(
+      EventDataValue eventDataValue,
+      Event event,
+      DataElement dataElement,
+      EntityManager entityManager,
+      TrackerPreheat preheat) {
+    if (dataElement.isFileType()) {
+      unassignFileResource(entityManager, preheat, event.getUid(), eventDataValue.getValue());
+    }
+
+    event.getEventDataValues().remove(eventDataValue);
   }
 
   private void logTrackedEntityDataValueHistory(
```

Comment:
```suggestion
  private void logDataValueChange(
```

or `addTrackedEntityDataValueChangeLog` as the service name (although I am unsure thats a good name).

Reply:
This adds a change log record for an event. I would prefer not to include the word `TrackedEntity` in the method name, as I believe it makes it difficult to distinguish from the attributes change log.

Reply:
definitely 👍🏻 

I guess this service method is just an old one with old names. The type itself is also `TrackedEntityDataValueChangeLog`.

The service is `EventChangeLogService`, how about
* `addDataValueChangeLog()`
* `add` might be good enough as well as you have the context from the service/parameter type on what changes you are tracking

Reply:
sounds good to me
there's [this ticket](https://dhis2.atlassian.net/browse/DHIS2-17835) that's about splitting change log from audit, we can tackle all these naming issues there